Forge/src/main/java/me/ferdz/placeableitems/client/renderer/tileentity/FluidHolderRenderer.java

```diff
@@ -0,0 +1,176 @@
+
+import me.ferdz.placeableitems.block.PlaceableItemsBlock;
+import me.ferdz.placeableitems.block.component.IBlockComponent;
+import me.ferdz.placeableitems.block.component.impl.FluidHolderBlockComponent;
+import me.ferdz.placeableitems.client.AllVertexBoundingBox;
+import me.ferdz.placeableitems.client.model.FluidModel;
+import me.ferdz.placeableitems.tileentity.FluidHolderTileEntity;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.BufferBuilder;
+import net.minecraft.client.renderer.Vector3d;
+import net.minecraft.client.renderer.texture.TextureAtlasSprite;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.util.Direction;
+import net.minecraft.util.math.BlockPos;
+
+import net.minecraftforge.client.model.animation.TileEntityRendererFast;
+import net.minecraftforge.fluids.FluidStack;
+
+/**
+ * A fast tile entity renderer implementation for {@link FluidHolderTileEntity} instances.
+ * Fluid holders should bind {@link FluidModel} implementations to each registered renderer
+ * instance using {@link #bind(PlaceableItemsBlock, FluidModel)}. If a
+ * {@link FluidHolderBlockComponent#shouldRenderFluid()} returns true but has no registeredFluidModel,
+ * an exception will be thrown and the client will crash.
+ *
+ * @author Parker Hawke - Choco
+ */
+public class FluidHolderRenderer extends TileEntityRendererFast<FluidHolderTileEntity> {
+
+    private final Map<PlaceableItemsBlock, FluidModel> fluidModels = new IdentityHashMap<>();
+
+    @Override
+    public void renderTileEntityFast(FluidHolderTileEntity tile, double x, double y, double z, float partialTicks, int destroyStage, BufferBuilder buffer) {
+        FluidStack fluidStack = tile.getFluid();
+        if (fluidStack.isEmpty()) {
+            return;
+        }
+
+        BlockState state = tile.getBlockState();
+        Block block = state.getBlock();
+        if (!(block instanceof PlaceableItemsBlock)) {
+            return;
+        }
+
+        FluidHolderBlockComponent fluidComponent = getFluidHolderBlockComponent((PlaceableItemsBlock) block);
+        Preconditions.checkState(fluidComponent != null, "FluidHolderTileEntity does not have a FluidHolderBlockComponent. This is impossible.");
+        if (!fluidComponent.shouldRenderFluid()) {
+            return;
+        }
+
+        FluidModel model = this.fluidModels.get(block);
+        Preconditions.checkState(model != null, "Missing model binding for FluidHolderTileEntity (" + tile.getClass().getName() + ")");
+
+        Vector3d rotationPoint = model.getRotationPoint(state);
+        buffer.setTranslation(x + (rotationPoint.x / 16F), y + (rotationPoint.y / 16F), z + (rotationPoint.z / 16F));
+
+        // Render the fluid
+        Fluid fluid = fluidStack.getFluid();
+        TextureAtlasSprite still = Minecraft.getInstance().getTextureMap().getAtlasSprite(fluid.getAttributes().getStill(fluidStack).toString());
```

Comment:
Does this support animations & mcmetas?

Reply:
Yes. The fluids are animated. This includes modded fluids as well.